source/simulator/assembly.cc

```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                ||
+                (!advection_field.is_temperature()))
+              {
+                //We are either in the case of a Dirichlet temperature boundary, or we are assembling for a composition variable. In the former, impose the Dirichlet value weakly using a matrix term and RHS term. In the latter, if this is an inflow boundary then impose a zero Dirichlet value by assembling the corresponding matrix term but no RHS term
+
+                for (unsigned int q=0; q<n_q_points; ++q)
+                  {
+                    // precompute the values of shape functions and their gradients.
+                    // We only need to look up values of shape functions if they
+                    // belong to 'our' component. They are zero otherwise anyway.
+                    // Note that we later only look at the values that we do set here.
+                    for (unsigned int k=0; k<advection_dofs_per_cell; ++k)
+                      {
+                        scratch.face_grad_phi_field[k]      = scratch.face_finite_element_values[solution_field].gradient (scratch.face_finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
+                        scratch.face_phi_field[k]      = scratch.face_finite_element_values[solution_field].value (scratch.face_finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
+                      }
+                    const double density_c_P              =
+                      ((advection_field.is_temperature())
+                       ?
+                       scratch.face_material_model_outputs.densities[q] *
+                       scratch.face_material_model_outputs.specific_heat[q]
+                       :
+                       1.0);
+
+                    Assert (density_c_P >= 0,
+                            ExcMessage ("The product of density and c_P needs to be a "
+                                        "non-negative quantity."));
+
+                    const double conductivity =
+                      ((advection_field.is_temperature())
+                       ?
+                       scratch.face_material_model_outputs.thermal_conductivities[q]
+                       :
+                       0.0);
+                    const double latent_heat_LHS =
+                      ((advection_field.is_temperature())
+                       ?
+                       face_heating_model_outputs.lhs_latent_heat_terms[q]
+                       :
+                       0.0);
+                    Assert (density_c_P + latent_heat_LHS >= 0,
+                            ExcMessage ("The sum of density times c_P and the latent heat contribution "
+                                        "to the left hand side needs to be a non-negative quantity."));
+
+                    const double penalty = (advection_field.is_temperature()
```

Comment:
Could you document where the magic 10 comes from? Also why is temperature and composition handled differently? Due to the thermal diffusion?